public/app/plugins/panel/gauge/module.tsx

```diff
@@ -1,10 +1,132 @@
+const optionsModel: OptionsUIModel<GaugeOptions> = {
+  model: {
+    type: OptionsUIType.Layout,
+    config: {
+      columns: 1,
+    },
+    content: [
+      {
+        type: OptionsUIType.Layout,
+        config: { columns: 3 },
+        content: [
+          {
+            type: OptionsUIType.Group,
+            config: { title: 'Thresholds' },
+            component: PanelOptionsGroup,
+            content: [
+              {
+                type: OptionsUIType.Editor,
+                editor: {
+                  optionType: OptionType.Object,
+                  component: ThresholdsEditor,
+                  property: 'thresholds',
+                },
+              } as OptionEditor<GaugeOptions, 'thresholds'>,
+            ],
+          } as OptionsPanelGroup,
+          {
+            type: OptionsUIType.Group,
+            config: { title: 'Value settings' },
+            component: PanelOptionsGroup,
+            content: [
+              {
+                type: OptionsUIType.Editor,
+                editor: {
+                  optionType: OptionType.Object,
+                  component: SingleStatValueEditor,
+                  property: 'valueOptions',
+                },
+              } as OptionEditor<GaugeOptions, 'valueOptions'>,
+            ],
+          } as OptionsPanelGroup,
+          {
+            type: OptionsUIType.Group,
+            config: { title: 'Gauge' },
+            component: PanelOptionsGroup,
+            content: [
+              {
+                type: OptionsUIType.Editor,
+                editor: {
+                  optionType: OptionType.Number,
+                  component: IntegerOption,
+                  property: 'minValue',
+                  label: 'Min value',
+                },
+              } as OptionEditor<GaugeOptions, 'minValue'>,
+              {
+                type: OptionsUIType.Editor,
+                editor: {
+                  optionType: OptionType.Number,
+                  component: IntegerOption,
+                  property: 'maxValue',
+                  label: 'Max value',
+                },
+              } as OptionEditor<GaugeOptions, 'maxValue'>,
+            ],
+          } as OptionsPanelGroup,
+        ],
+      } as OptionsGrid,
+    ],
+  } as OptionsGrid,
+};
+
+const valueOptionsSchema: ObjectOptionDataSchema<SingleStatBaseOptions> = {
+  properties: {
+    orientation: {},
+    thresholds: {},
+    valueMappings: {},
+    valueOptions: {},
+  },
+};
+
+const optionsSchema: OptionsDataSchema<GaugeOptions> = {
+  title: 'GaugeOptions',
+  type: 'object',
+  required: ['minValue', 'maxValue'],
+  properties: {
+    minValue: {
+      type: 'number',
+      description: 'Hint for min value...',
+    },
+    maxValue: {
+      type: 'number',
+      description: 'Hint for min value...',
+    },
+    showThresholdMarkers: {
+      type: 'boolean',
+    },
+    showThresholdLabels: {
+      type: 'boolean',
+    },
+    ...valueOptionsSchema.properties,
```

Comment:
For the schema to be more useful we would need to extend it to include the higher level concepts like type: "unit", type: "valueOptions", intead of just the raw bools & strings.

Reply:
I sort of agree. I think things like i.e. value options are in the end of `object` type with nested primitives. For data schema that will be helpful for validation I found Yup used by Formik for instance (https://github.com/jquense/yup) a nice way for data schema representation and validation. I haven't found similar API for JSON Schema though, that would help us build data schema without much effort.

Both approaches will enable us to to have those more abstract schemas, actually in this PR I have already naively tried that with: 

```ts
const valueOptionsSchema: ObjectOptionDataSchema<SingleStatBaseOptions> = {
  properties: {
    orientation: {},
    thresholds: {},
    valueMappings: {},
    valueOptions: {},
  },
};
```

that's ltr on composed in Gauge options schema:
```ts
  // ...
  properties: {
    minValue: {
      type: 'number',
      description: 'Hint for min value...',
    },
    maxValue: {
      type: 'number',
      description: 'Hint for min value...',
    },
    showThresholdMarkers: {
      type: 'boolean',
    },
    showThresholdLabels: {
      type: 'boolean',
    },
    ...valueOptionsSchema.properties,
  },
 // ...
```

In this case value options are not nested in Gauge options, which I think could be useful, especially when there are cases with lots of options that could be grouped (see react graph panel for reference https://github.com/grafana/grafana/blob/master/public/app/plugins/panel/graph2/types.ts#L14)

There are of course JSON Schema validators that we could use for data validation (i.e. https://ajv.js.org/) but I haven't found anything close to Yup when it comes to how easy it is to both create and enable schema validation.